Give a typical example of Makefile wrapper for coq_makefile (addresses #12903) (#13220)
coq/coq: doc/sphinx/practical-tools/utilities.rst

  is a makefile for ``GNU Make`` with targets to build the project
  (e.g. generate .vo or .html files from .v or compile .ml* files)
  and install it in the ``user-contrib`` directory where the |Coq|
  library is installed.

CoqMakefile.conf
  contains make variables assignments that reflect
  the contents of the ``_CoqProject`` file as well as the path relevant to
  |Coq|.

The recommended approach is to wrap ``CoqMakefile`` into a standard ``Makefile`` of the form:

::

    all: CoqMakefile
            +make -f CoqMakefile all

    clean: CoqMakefile
            +make -f CoqMakefile clean

>             rm -f Makefile.coq

Reviewer: ```suggestion
            rm -f CoqMakefile
```
?
Why CoqMakefile anyway, I've never seen it used unlike Makefile.coq?

    CoqMakefile: _CoqProject
            $(COQBIN)coq_makefile -f _CoqProject -o CoqMakefile

    _CoqProject: ;

    %: CoqMakefile

>             +make -f CoqMakefile $@

Reviewer: Should we show a mechanism like https://github.com/coq/coq/blob/e583be62b74d71b5af159700e3a31f78fec9a7d2/Makefile.make#L166-L178 so that `make a b` results in a single subcall to make? Also
```suggestion
            $(MAKE) -f CoqMakefile $@
```
is the recommended way.

``Makefile`` of the following form:

.. example::

  ::

      # KNOWNTARGETS will not be passed along to CoqMakefile
      KNOWNTARGETS := CoqMakefile extra-stuff extra-stuff2
      # KNOWNFILES will not get implicit targets from the final rule, and so
      # depending on them won't invoke the submake
      # Warning: These files get declared as PHONY, so any targets depending
      # on them always get rebuilt
      KNOWNFILES   := Makefile _CoqProject

      .DEFAULT_GOAL := invoke-coqmakefile

      CoqMakefile: Makefile _CoqProject
              $(COQBIN)coq_makefile -f _CoqProject -o CoqMakefile

      invoke-coqmakefile: CoqMakefile
              $(MAKE) --no-print-directory -f CoqMakefile $(filter-out $(KNOWNTARGETS),$(MAKECMDGOALS))

      .PHONY: invoke-coqmakefile $(KNOWNFILES)

      ####################################################################
      ##                      Your targets here                         ##
      ####################################################################

      # This should be the last rule, to handle any targets not declared above
      %: invoke-coqmakefile
              @true

> The advantage of a wrapper, compared to directly calling the generated
> ``Makefile`` or to including it with an include directive, is that it
> provides a target independent of the version of Coq to regenerate a
> ``Makefile`` specific to the current version of Coq. Additionally, the
> master ``Makefile`` can be extended with targets not specific to Coq.

Reviewer 2: ```suggestion
The advantage of a wrapper, compared to directly calling the generated
``Makefile``, is that it
provides a target independent of the version of Coq to regenerate a
``Makefile`` specific to the current version of Coq. Additionally, the
master ``Makefile`` can be extended with targets not specific to Coq.
Including the generated makefile with an include directive is
discouraged, snice the contents of this file, including variable names and
status of rules, may change in the future.
```

Author: For me this is ideal. Let me already commit the suggestion.

  is a makefile for ``GNU Make`` with targets to build the project
  (e.g. generate .vo or .html files from .v or compile .ml* files)
  and install it in the ``user-contrib`` directory where the |Coq|
  library is installed.

CoqMakefile.conf
  contains make variables assignments that reflect
  the contents of the ``_CoqProject`` file as well as the path relevant to
  |Coq|.

> The recommended approach is to embed ``CoqMakefile`` into a standard

Reviewer 2: ```suggestion
The recommended approach is to invoke ``CoqMakefile`` from a standard
```

Reviewer 2: Also, I want to keep the sentence that says that including the generated makefile is discouraged.

.. example::

  ::

      # KNOWNTARGETS will not be passed along to CoqMakefile
      KNOWNTARGETS := CoqMakefile extra-stuff extra-stuff2
      # KNOWNFILES will not get implicit targets from the final rule, and so
      # depending on them won't invoke the submake
      # Warning: These files get declared as PHONY, so any targets depending
      # on them always get rebuilt
      KNOWNFILES   := Makefile _CoqProject

      .DEFAULT_GOAL := invoke-coqmakefile

      CoqMakefile: Makefile _CoqProject
              $(COQBIN)coq_makefile -f _CoqProject -o CoqMakefile

      invoke-coqmakefile: CoqMakefile
              $(MAKE) --no-print-directory -f CoqMakefile $(filter-out $(KNOWNTARGETS),$(MAKECMDGOALS))

      .PHONY: invoke-coqmakefile $(KNOWNFILES)

      ####################################################################
      ##                      Your targets here                         ##
      ####################################################################

      # This should be the last rule, to handle any targets not declared above
      %: invoke-coqmakefile
              @true

An alternative would have been to give instead option ``-o Makefile``

> to ``coq_makefile`` but the advantage of a wrapper is that it provides

Reviewer 3: I understand everything starting with "the advantage of a wrapper" but nothing before. It would read fine if the sentence were just to start at this point.

Author: OK, that's right that what comes before is in some sense implicit.

a way independent of the version of Coq to regenerate a ``Makefile``
specific to the current version of Coq. Additionally, the master
``Makefile`` can be extended with targets not specific to Coq.

An optional file ``CoqMakefile.local`` can be provided by the user in order to

> extend ``CoqMakefile`` with other Coq targets.

Reviewer 2: This is not correct. If you want to add a brand new target, you can just add it your wrapping makefile.
Makefile.local is to extend existing targets or set variables.

Author: I was rephrasing the text that followed: "Similarly one can customize the install target or even provide new targets." Does this sentence need to be changed also?

Reviewer 2: OK, let me rephrase: "Makefile.local is mainly to extend existing targets or set variables. You can also add new targets".
The current text does not even mention what I consider to be the main use cases.

Author: @gares (and @JasonGross): I pushed an update but I'm not fully sure I expressed correctly what you intended. Don't hesitate to make a suggestion.

      This target requires python to build the table.

Reusing/extending the generated Makefile
++++++++++++++++++++++++++++++++++++++++

> Including the generated makefile with an include directive is

Reviewer 2: I want to keep this sentence, please don't remove it.

Author: Could this sentence be embedded in a previous sentence, as in:
```
the advantage of a wrapper (compared e.g. to including the generated makefile with an include directive) is that it provides a way independent of the version of Coq to regenerate a ``Makefile``
```
?

Reviewer 4: One drawback is that you can't have your own target for generating the included Makefile, because the generated Makefile includes a target regenerating itself, and there's not a good way to override that target.

Author: So:
```
... the advantage of a wrapper (compared to directly calling the generated makefile
or to including the generated makefile with an include directive) is that it provides a way 
independent of the version of Coq to regenerate a ``Makefile`` ...
```
?

Reviewer 2: Saying that A has advantages over B is not the same thing as saying that you should no do B.
The reasons are other, like that `inclusion` is not modular, all variables and rules are injected into the same name space.
This is why there is Makefile.conf that only include variables assignment and all variables are named COQMF_*, that one you can include.
